core/thread/token.go

```diff
@@ -0,0 +1,189 @@
+package thread
+
+import (
+	"context"
+	"encoding/base64"
+	"fmt"
+	"log"
+	"strings"
+	"time"
+
+	"github.com/dgrijalva/jwt-go"
+	"github.com/gogo/status"
+	"github.com/grpc-ecosystem/go-grpc-middleware/util/metautils"
+	"github.com/libp2p/go-libp2p-core/crypto"
+	jwted25519 "github.com/textileio/go-threads/jwt"
+	"google.golang.org/grpc/codes"
+)
+
+// Identity represents an entity capable of signing a message
+// and returning the associated public key for verification.
+// In many cases, this will just be a private key, but callers
+// can use any setup that suits their needs.
+type Identity interface {
+	Sign([]byte) ([]byte, error)
+	GetPublic() PubKey
+}
+
+// Libp2pIdentity wraps crypto.PrivKey, overwriting GetPublic with thread.PubKey.
+type Libp2pIdentity struct {
+	crypto.PrivKey
+}
+
+// NewLibp2pIdentity returns a new Libp2pIdentity.
+func NewLibp2pIdentity(key crypto.PrivKey) Identity {
+	return &Libp2pIdentity{PrivKey: key}
+}
+
+func (p *Libp2pIdentity) GetPublic() PubKey {
+	return NewLibp2pPubKey(p.PrivKey.GetPublic())
+}
+
+// Pubkey can be anything that provides a verify method.
+type PubKey interface {
```

Comment:
This is meant to allow for all kinds of pub keys, with an implementation for libp2p pubkeys below. This also has built in string methods that should encode / decode with base64 for use in a JWT.